### PR TITLE
Add support for extracting databricks dependencies for various retriever types

### PR DESCRIPTION
**Pull request by annzhang-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/annzhang-db/mlflow/pull/12722?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12722/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12722
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

- Add casing in `_extract_databricks_dependencies_from_retriever` for more types of retrievers from langchain.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests
   - only manual tested for EnsembleRetriever so far, planning to do manual tests for other types

Screenshot 2024-07-19 at 3 37 41 PM
Screenshot 2024-07-19 at 3 39 13 PM

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/langchain/databricks_dependencies.py [hunk 1]
```diff
@@ -48,6 +48,23 @@ def _extract_databricks_dependencies_from_retriever(retriever) -> Generator[Reso
...
 
+def _extract_databricks_dependencies_from_retriever(retriever) -> Generator[Resource, None, None]:
+    if hasattr(retriever, "base_retriever"):
+        retriever = getattr(retriever, "base_retriever", None)
+
+    if hasattr(retriever, "retriever"):
+        retriever = getattr(retriever, "retriever", None)
+
+    if hasattr(retriever, "retrievers"):
+        retriever = getattr(retriever, "retrievers", None)
```
Comment (prithvikannan) [inline mlflow/langchain/databricks_dependencies.py:59]: non blocking but curious why these are needed?

File: tests/langchain/test_langchain_databricks_dependency_extraction.py [hunk 2]
```diff
@@ -134,11 +137,69 @@ def test_parsing_dependency_from_databricks_retriever(monkeypatch: pytest.Monkey
...
+    from langchain.retrievers.document_compressors import LLMChainExtractor
+
+    compressor = LLMChainExtractor.from_llm(llm)
+    compression_retriever = ContextualCompressionRetriever(
+        base_compressor=compressor, base_retriever=retriever_1
+    )
+    assert list(_extract_databricks_dependencies_from_retriever(compression_retriever)) == [
+        DatabricksVectorSearchIndex(index_name="mlflow.rag.vs_index_1"),
+        DatabricksServingEndpoint(endpoint_name="databricks-bge-large-en"),
+    ]
```
Comment (prithvikannan) [inline tests/langchain/test_langchain_databricks_dependency_extraction.py:182]: this appears to be using ChatOpenAI as llm. if we use ChatDatabricks in here, would we pick up that dependency too?

apologies if im missing something here.

**Review by prithvikannan — APPROVED** [review]
LGTM! Thanks for adding this support, and for the thorough testing

File: mlflow/langchain/databricks_dependencies.py (hunk 1 above)
Comment (B-Step62) [reply mlflow/langchain/databricks_dependencies.py:59]: +1, could you add comments or docstring explaining for which case we need this extracting? While we can find some clues in LangChain code base, as their code base changes rapidly, it is often helpful to have some note in our code base🙂

File: tests/langchain/test_langchain_databricks_dependency_extraction.py [hunk 3]
```diff
@@ -123,7 +123,10 @@ def test_parsing_dependency_from_databricks_retriever(monkeypatch: pytest.Monkey
     from langchain.vectorstores import DatabricksVectorSearch
```
Comment (B-Step62) [inline tests/langchain/test_langchain_databricks_dependency_extraction.py:123]: We have two tests with exactly same name `test_parsing_dependency_from_databricks_retriever`. Pytest only runs one of them. Apparently the test contents are exactly same aside from the minor import path difference. Can you remove the duplication?

**Comment by B-Step62** [comment]
Merging this to include in the RC cut, @annzhang-db would be awesome if you could address ^ as a follow-up!

**Pull request merged**